plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts

```diff
@@ -127,3 +129,130 @@ describe('TaskWorker', () => {
     expect(event?.body.output).toEqual({ testOutput: 'testmockoutput' });
   });
 });
+
+describe('Concurrent TaskWorker', () => {
+  let storage: DatabaseTaskStore;
+
+  const integrations: ScmIntegrations = {} as ScmIntegrations;
+
+  const actionRegistry: TemplateActionRegistry = {} as TemplateActionRegistry;
+  const workingDirectory = os.tmpdir();
+  let asyncTasksCount = 0;
+
+  const workflowRunner: NunjucksWorkflowRunner = {
+    execute: () => {
+      asyncTasksCount++;
+      return new Promise(resolve => {
+        setTimeout(() => {
+          resolve({ output: { testOutput: 'testmockoutput' } });
+        }, 1000);
+      });
+    },
+  } as unknown as NunjucksWorkflowRunner;
+
+  beforeAll(async () => {
+    storage = await createStore();
+  });
+
+  beforeEach(() => {
+    asyncTasksCount = 0;
+    jest.resetAllMocks();
+    MockedNunjucksWorkflowRunner.mockImplementation(() => workflowRunner);
+  });
+
+  const logger = getVoidLogger();
+
+  it('should be able to run multiple tasks at once', async () => {
+    const broker = new StorageTaskBroker(storage, logger);
+
+    const dispatchANewTask = () =>
+      broker.dispatch({
+        spec: {
+          apiVersion: 'scaffolder.backstage.io/v1beta3',
+          steps: [{ id: 'test', name: 'test', action: 'not-found-action' }],
+          output: {
+            result: '{{ steps.test.output.testOutput }}',
+          },
+          parameters: {},
+        },
+      });
+
+    const expectedConcurrentTasks = 3;
+    const taskWorker = await TaskWorker.create({
+      logger,
+      workingDirectory,
+      integrations,
+      taskBroker: broker,
+      actionRegistry,
+      concurrentTasksLimit: expectedConcurrentTasks,
+    });
+
+    taskWorker.start();
+
+    await dispatchANewTask();
+    await dispatchANewTask();
+    await dispatchANewTask();
+    await dispatchANewTask();
+
+    expect(asyncTasksCount).toEqual(expectedConcurrentTasks);
+  });
+});
+
+describe('TaskWorker internals', () => {
+  const TaskWorkerConstructor = TaskWorker as unknown as {
+    new (options: TaskWorkerOptions): TaskWorker;
+  };
+
+  it('should not pick up tasks before it is ready to execute more work', async () => {
```

Comment:
What I mentioned in https://github.com/backstage/backstage/pull/13817/files#r998666872 still seems to be an issue, so I added this failing test to highlight what I mean. Does this clarify my concern?

Reply:
Thanks @Rugvip I changed how the queue claims tasks (only when the number of pending tasks is less than the limit will the worker claim a task). Your test now passes.